contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -441,37 +441,112 @@ Alternatively, deployment could be simplified by having all components (includin
 
 ### CSI Credentials
 
-This part of proposal is not going to be implemented in alpha release.
+CSI allows specifying credentials in CreateVolume/DeleteVolume, ControllerPublishVolume/ControllerUnpublishVolume, and NodePublishVolume/NodeUnpublishVolume operations.
 
-#### End user credentials
-CSI allows specifying *end user credentials* in all operations. Kubernetes does not have facility to configure a Secret per *user*, we usually track objects per *namespace*. Therefore we decided to postpone implementation of these credentials and wait until CSI is clarified.
+Kubernetes will enable cluster admins and users deploying workloads on the cluster to specify these credentials by referencing Kubernetes secret object(s). Kubernetes (either the core components or helper containers) will fetch the secret(s) and pass them to the CSI volume plugin.
 
-#### Volume specific credentials
-Some storage technologies (e.g. iSCSI with CHAP) require credentials tied to the volume (iSCSI LUN) that must be used during `NodePublish` request. It is expected that these credentials will be provided during dynamic provisioning of the volume, however CSI `CreateVolume` response does not provide any. In case it gets fixed soon external provisioner can save the secrets in a dedicated namespace and make them available to external attacher and internal CSI volume plugin using these `CSIPersistentVolumeSource` fields:
+If a secret object contains more than one secret, all secrets are passed.
+
+#### Secret to CSI Credential Encoding
+
+CSI accepts credentials for all the operations specified above as a map of string to string (e.g. `map<string, string> controller_create_credentials`).
+
+Kubernetes, however, defines secrets as a map of string to byte-array (e.g. `Data map[string][]byte`). It also allows specifying text secret data in string form via a write-only convenience field `StringData` which is a map of string to string.
+
+Kubernetes also requires secret string data to be base64 encoded, but CSI does not.
+
+Therefore, before passing secret data to CSI, Kubernetes (either the core components or helper containers) will convert the secret data from bytes to string (Kubernetes does not specify the character encoding, but Kubernetes internally uses golang to cast from string to byte and vice versa which assumes UTF-8 character set), and the resulting string will be run through base64 decoding to get the plain secret to pass to CSI.
```

Comment:
> the resulting string will be run through base64 decoding to get the plain secret to pass to CSI

I'm unsure what this means... an example might be useful

Reply:
https://kubernetes.io/docs/concepts/configuration/secret/#creating-a-secret-manually says `Each item must be base64 encoded`. No such restriction exists on the CSI side (a plain string is allowed).

Reply:
That documentation only applies to the JSON API. The secret must be base64-encoded when submitted as json because it is allowing for storing binary data in a json string.

If CSI only supports string data, then we just need to define what happens when a referenced secret contains binary data that cannot be represented in a CSI string.

Reply:
hmm, maybe I'm importing the problems we had with json-serialized golang strings... 

from https://blog.golang.org/strings:

> It's important to state right up front that a string holds arbitrary bytes. It is not required to hold Unicode text, UTF-8 text, or any other predefined format. As far as the content of a string is concerned, it is exactly equivalent to a slice of bytes.

As long as all the CSI API serializations and clients hold the same view of strings (arbitrary bytes, not assumed to be UTF-8), then secret `map[string][]byte` -> CSI `map[string]string` is lossless.

Reply:
Ack. In that case conversion from byte to string should be all that is needed. I will drop requirement for base64 decode.